{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Collecting pysam\n",
      "  Downloading pysam-0.16.0.1-cp38-cp38-manylinux1_x86_64.whl (10.2 MB)\n",
      "\u001b[K     |████████████████████████████████| 10.2 MB 2.8 MB/s eta 0:00:01\n",
      "\u001b[?25hInstalling collected packages: pysam\n",
      "Successfully installed pysam-0.16.0.1\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install pysam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pysam as pys\n",
    "from pysam import VariantFile"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "class mutations_per_chromozom:\n",
    "    def __init__(self, chromozom, positions):\n",
    "        self.chromozom = chromozom\n",
    "        self.positions = positions\n",
    "    \n",
    "\n",
    "def extract_predictions_from_VCF (file_name):\n",
    "    num = 0\n",
    "    array = [] # [1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22, X (23), Y (24), MT (25)]\n",
    "    handle = VariantFile(file_name)\n",
    "    index  = 1\n",
    "    \n",
    "    array.append(0)\n",
    "    for i in range (1,26):\n",
    "        array.append(mutations_per_chromozom(i,[]))\n",
    "    \n",
    "    # Calculate only SNPs\n",
    "    for record in handle.fetch():\n",
    "        allels = record.alleles\n",
    "        if (len(allels) == 2):\n",
    "            ref, alt = record.alleles\n",
    "            if(len(ref) == 1 and len(alt) == 1): \n",
    "                num = num + 1\n",
    "                if (record.chrom == 'X'):\n",
    "                    array[23].positions.append(record.pos)\n",
    "                elif(record.chrom == 'Y'):\n",
    "                    array[24].positions.append(record.pos)\n",
    "                elif(record.chrom == 'MT'):\n",
    "                    array[25].positions.append(record.pos)\n",
    "                else:\n",
    "                     array[int(record.chrom)].positions.append(record.pos)\n",
    "\n",
    "    # returning predictions number and all prediction posssitions in array.\n",
    "    return num, array \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of predictions in VCF truth set is:  62295\n"
     ]
    }
   ],
   "source": [
    "# count number of predictions in VCF truth set (gatk_haplotype_caller)\n",
    "\n",
    "gatk_vcf = \"/sbgenomics/workspace/output.vcf\"\n",
    "gatk_predictions_num, gatk_predictions_array = extract_predictions_from_VCF(gatk_vcf)  \n",
    "print (\"Number of predictions in VCF truth set is: \",gatk_predictions_num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of predictions in VCF truth set is:  146717\n"
     ]
    }
   ],
   "source": [
    "# count number of predictions in VCF test set (freebayes)\n",
    "\n",
    "freebayes_vcf = \"/sbgenomics/workspace/var.vcf\"\n",
    "freebayes_predictions_num, freebayes_predictions_array = extract_predictions_from_VCF(freebayes_vcf)  \n",
    "print (\"Number of predictions in VCF truth set is: \", freebayes_predictions_num)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of TRUE POSITIVES  :=  58578\n",
      "Number of FALSE POSITIVES :=  88139\n"
     ]
    }
   ],
   "source": [
    "# Calculating number of True/False positives.\n",
    "# VCF generated by GATK is truth set and VCF generated by freebayes is test set\n",
    "# True positive is number of predictions which exist in both VCF files.\n",
    "# Freebayes is test set, so we take prediction from it and then check do that prediction exist in GATK VCF.\n",
    "# If prediction (mutation) exist in gatk that is TRUE POSITIVE, otherwise is FALSE POSITIVE\n",
    "\n",
    "# We need to callculate those values for each chromozom and then we will add them.\n",
    "\n",
    "def calculate_true_and_false_positive_values (truth_set_array, test_set_array):\n",
    "    true_positive = 0\n",
    "    false_positive = 0\n",
    "    for chrom in range (1,26): # 1..25\n",
    "        for position in test_set_array[chrom].positions:\n",
    "            if position in truth_set_array[chrom].positions:\n",
    "                true_positive = true_positive + 1\n",
    "            else:\n",
    "                false_positive = false_positive + 1\n",
    "     \n",
    "    return true_positive, false_positive\n",
    "\n",
    "true, false = calculate_true_and_false_positive_values(gatk_predictions_array, freebayes_predictions_array)\n",
    "\n",
    "print (\"Number of TRUE POSITIVES  := \", true)\n",
    "print (\"Number of FALSE POSITIVES := \", false)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "TRUE_POSITIVES + FALSE_POSITIVES :=  146717\n"
     ]
    }
   ],
   "source": [
    "# We can see that number of TRUE_POSITIVES is just a little bit lower from number of predictions in GATK\n",
    "# Important thing here is that number of TRUE_POSITIVES + FALSE_POSITIVES is exactly the same  as number \n",
    "# of predictions in freebayes VCF file.\n",
    "\n",
    "print(\"TRUE_POSITIVES + FALSE_POSITIVES := \", true + false )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FALSE_NEGATIVES :=  3717\n"
     ]
    }
   ],
   "source": [
    "# On similar way we can calculate number of FALSE NEGATIVES\n",
    "# as number of predictions which are existing in GATK but not in FREEBAYES\n",
    "\n",
    "def calculate_false_negative_values (truth_set_array, test_set_array):\n",
    "    false_negative = 0\n",
    "    for chrom in range (1,26): # 1..25\n",
    "        for position in truth_set_array[chrom].positions:\n",
    "            if not position in test_set_array[chrom].positions:\n",
    "                false_negative = false_negative + 1\n",
    "     \n",
    "    return false_negative\n",
    "\n",
    "false_neg = calculate_false_negative_values (gatk_predictions_array, freebayes_predictions_array)\n",
    "\n",
    "print (\"FALSE_NEGATIVES := \", false_neg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of FALSE_NEGATIVES + TRUE POSITIVES :=  62295\n"
     ]
    }
   ],
   "source": [
    "# One thing to note is that sum of FALSE_NEGATIVES and TRUE POSITIVES is same as number of predictions in GATK VCF\n",
    "\n",
    "print (\"Number of FALSE_NEGATIVES + TRUE POSITIVES := \", true + false_neg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ALL CALCULATED DATA (SNPs only): \n",
      "\n",
      "Number of TRUE POSITIVES  :=  58578\n",
      "Number of FALSE POSITIVES :=  88139\n",
      "Number of FALSE_NEGATIVES :=  3717\n"
     ]
    }
   ],
   "source": [
    "# ALL CALCULATED DATA:\n",
    "\n",
    "print (\"ALL CALCULATED DATA (SNPs only): \\n\")\n",
    "print (\"Number of TRUE POSITIVES  := \", true)\n",
    "print (\"Number of FALSE POSITIVES := \", false)\n",
    "print (\"Number of FALSE_NEGATIVES := \", false_neg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "## DATA GENERATED BY TOOL\n",
    "## METRICS CLASS    picard.vcf.GenotypeConcordanceContingencyMetricsVARIANT_TYPE    \n",
    "##      TRUTH_SAMPLE    CALL_SAMPLE    TP_COUNT    TN_COUNT    FP_COUNT    FN_COUNT    EMPTY_COUNT\n",
    "## SNP    HCC1143        HCC1143        58372       28546       6620        2795           3\n",
    "## INDEL  HCC1143        HCC1143        3722        4851        5033        4163           52"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
